File: js/216_CombinationSumIII.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
    var ret = [];
    backtrace(ret,[],1,n,k);
    return ret;
};

var backtrace = function(ret,path,start,n,k){
    if(n === 0 && path.length === k){
        ret.push(path.concat());
        return;
    }
    if(n < 0) return;
    for(var i = start; i <= 9; ++i){
        path.push(i);
        backtrace(ret,path,i+1,n-i,k);
        path.pop();
    }
}

File: cpp/216_CombinationSumIII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ret;
        vector<int> path;
        backtrack(ret,path,1,n,k);
        return ret;
    }

private:
    void backtrack(vector<vector<int>>& ret, vector<int>& path, int start,int n, int k){
        if(n == 0 && path.size() == k){
            ret.push_back(path);
            return;
        }
        if(n < 0) return;
        for(int i = start; i <= 9; ++i){
            path.push_back(i);
            backtrack(ret,path,i+1,n-i,k);
            path.pop_back();
        }
    }
};